tools/cluster-checker/checker.js

```diff
@@ -0,0 +1,247 @@
+'use strict'
+
+const k8s = require('@kubernetes/client-node')
+
+class Client {
+  constructor () {
+    const config = new k8s.KubeConfig()
+    config.loadFromDefault()
+    config.getCurrentCluster().skipTLSVerify = true
+    this.core = config.makeApiClient(k8s.CoreV1Api)
+    this.custom = config.makeApiClient(k8s.CustomObjectsApi)
+  }
+
+  async nodes () {
+    const res = await this.core.listNode()
+    return res.body.items
+  }
+
+  async namespaces () {
+    const res = await this.core.listNamespace()
+    return res.body.items
+  }
+
+  async pods (namespace) {
+    const res = await this.core.listNamespacedPod(namespace)
+    return res.body.items
+  }
+
+  async clusterQueues () {
+    const res = await this.custom.listClusterCustomObject(
+      'kueue.x-k8s.io',
+      'v1beta1',
+      'clusterqueues')
+    return res.body.items
+  }
+
+  async localQueues (namespace) {
+    const res = await this.custom.listNamespacedCustomObject(
+      'kueue.x-k8s.io',
+      'v1beta1',
+      namespace,
+      'localqueues')
+    return res.body.items
+  }
+
+  async workloads (namespace) {
+    const res = await this.custom.listNamespacedCustomObject(
+      'kueue.x-k8s.io',
+      'v1beta1',
+      namespace,
+      'workloads')
+    return res.body.items
+  }
+}
+
+// return the number of GPUs reserved by the pod
+function reservation (pod) {
+  if (pod.status?.phase === 'Succeeded' || pod.status?.phase === 'Failed') {
+    return 0 // pod has already completed
+  }
+  let scheduled = false
+  for (const condition of pod.status?.conditions ?? []) {
+    if (condition.type === 'PodScheduled' && condition.status === 'True') {
+      scheduled = true
+      break
+    }
+  }
+  if (!scheduled) {
+    return 0 // pod has not been scheduled yet
+  }
+  let gpus = 0
+  // compute sum of regular containers
+  for (const container of pod.spec.containers) {
+    if (container.resources?.requests?.['nvidia.com/gpu'] > 0) {
+      gpus += parseInt(container.resources?.requests?.['nvidia.com/gpu'])
+    } else if (container.resources?.limits?.['nvidia.com/gpu'] > 0) {
+      gpus += parseInt(container.resources?.limits?.['nvidia.com/gpu'])
+    }
+  }
+  // compute max with init containers
+  for (const container of pod.spec.initContainers ?? []) {
```

Comment:
init containers run in parallel.  Don't you need to sum their GPU usage together and then do the max operation outside the loop?

Reply:
This is not my understanding of init containers. See here: https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#understanding-init-containers.

Reply:
So much for trusting my memory... I really thought we had multiple init containers somewhere in the past in KAR that ran in parallel... comment withdrawn.